File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/internal/validation/RequiredParametersValidator.java
package com.emerchantpay.gateway.genesisandroid.api.internal.validation;

import com.emerchantpay.gateway.genesisandroid.api.constants.ErrorMessages;
import com.emerchantpay.gateway.genesisandroid.api.models.GenesisError;

import java.util.ArrayList;
import java.util.HashMap;

public class RequiredParametersValidator extends GenesisValidator {
    private HashMap<String, String> parametersMap;
    private ArrayList<String> missingParamsList = new ArrayList<String>();

    // Genesis error handler
    private GenesisError error;

    public RequiredParametersValidator() {
        super();
    }

    public RequiredParametersValidator(HashMap<String, String> parametersMap) {
        super();

        this.parametersMap = parametersMap;
    }

    public Boolean isValidRequiredParams() {
        // Missing params
        StringBuilder missingParams = new StringBuilder();

        for (String key : parametersMap.keySet()) {
            if (parametersMap.get(key) == null || parametersMap.get(key).isEmpty()) {
                missingParamsList.add(key);
            }
        }

        for (Integer i = 0; i < missingParamsList.size(); i++) {
            if (i == 0) {
                missingParams.append(missingParamsList.get(0));
            } else {
                missingParams.append(", " + missingParamsList.get(i));
            }
        }

        if (missingParamsList.size() > 1) {
            error = new GenesisError(ErrorMessages.REQUIRED_PARAMS, String.valueOf(missingParams));
            return false;
        } else if (!missingParamsList.isEmpty()) {
            error = new GenesisError(ErrorMessages.REQUIRED_PARAM, String.valueOf(missingParams));
            return false;
        } else {
            return true;
        }
    }

    public GenesisError getError() {
        return error;
    }

    public HashMap<String, String> setParameters(HashMap<String, String> parameters) {
        this.parametersMap = parameters;
        return this.parametersMap;
    }
}

File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/internal/validation/RequiredParametersValidator.kt
package com.emerchantpay.gateway.genesisandroid.api.internal.validation

import com.emerchantpay.gateway.genesisandroid.api.constants.ErrorMessages
import com.emerchantpay.gateway.genesisandroid.api.models.GenesisError
import java.util.*

class RequiredParametersValidator : GenesisValidator {
    private var parametersMap: HashMap<String, String>? = null
    private val missingParamsList = ArrayList<String>()

    // Genesis error handler
    override var error: GenesisError? = null
        set(value: GenesisError?) {
            super.error = value
        }

    // Missing params
    val isValidRequiredParams: Boolean?
        get() {
            val missingParams = StringBuilder()

            parametersMap!!.keys.forEach { key ->
                when {
                    parametersMap!![key] == null || parametersMap!![key]!!.isEmpty() -> missingParamsList.add(key)
                }
            }

            missingParamsList.indices.forEach { i ->
                when (i) {
                    0 -> missingParams.append(missingParamsList[0])
                    else -> missingParams.append(", " + missingParamsList[i])
                }
            }

            return when {
                missingParamsList.size > 1 -> {
                    error = GenesisError(ErrorMessages.REQUIRED_PARAMS, missingParams.toString())
                    false
                }
                !missingParamsList.isEmpty() -> {
                    error = GenesisError(ErrorMessages.REQUIRED_PARAM, missingParams.toString())
                    false
                }
                else -> true
            }
        }

    constructor() : super() {}

    constructor(parametersMap: HashMap<String, String>?) : super() {
        this.parametersMap = parametersMap
    }

    fun setParameters(parameters: HashMap<String, String>?): HashMap<String, String>? {
        this.parametersMap = parameters
        return this.parametersMap
    }
}
